File: java/Solution206.java
package leetCode.java;

//LinkedList O(n) O(1)
class Solution206 {
    public ListNode reverseList(ListNode head) {
        ListNode pre = null;//直接設null, 不用建立dummy
        ListNode cur = head;
        while(cur != null){
            ListNode next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }
        return pre;//此時cur已經null
    }
}
/*
null    1 -> 2 -> null
^p      ^c

null <- 1    2 -> null
        ^p   ^c

null <- 1 <- 2 -> null
             ^p   ^c
*/

File: c++/Solution206.cpp
//LinkList O(n) O(1)
class Solution206 {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* cur = head;
        ListNode* pre = nullptr;
        while (cur) {
            ListNode* next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }
        return pre;
    }
};


//LinkList O(n) O(n)
class Solution206_2 {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next)
            return head;

        ListNode* last = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return last;
    }
};
/* Follow up 是用recursively做(效能較差)
vhead
1 -> 2 -> 3 -> 4 -> null

ListNode* last = reverseList(head->next);後
vhead
1 -> 2 <- 3 <- 4
               ^last

head->next->next = head;後
vhead
1 <- 2 <- 3 <- 4
               ^last

head->next = nullptr;後
        vhead
null <- 1 <- 2 <- 3 <- 4
                       ^last
*/